projects/angular/src/utils/list-key-navigation/key-navigation-list.ts

```diff
@@ -0,0 +1,133 @@
+ */
+
+import { AfterViewInit, Directive, ElementRef, Inject, Input, OnDestroy, Renderer2 } from '@angular/core';
+import {
+  focusElement,
+  getFlattenedFocusableItems,
+  initializeKeyListItems,
+  KeyNavigationCode,
+  setActiveKeyListItem,
+  validKeyNavigationCode,
+} from '@cds/core/internal';
+
+import { FocusableElementsToken } from './providers/focusable-items';
+import { ClrListKeyNavigationConfigToken } from './providers/key-navigation-list.config';
+import { getNextKeyListItem } from './utils';
+
+export interface KeyNavigationListConfig {
+  keyListItems: string;
+  layout: 'both' | 'horizontal' | 'vertical';
+  manageFocus: boolean;
+  manageTabindex: boolean;
+  loop: boolean;
+  dir: string | null;
+}
+
+@Directive({
+  selector: '[clrKeyNavigationList]',
+})
+export class ClrKeyNavigationList implements AfterViewInit, OnDestroy {
+  private config: KeyNavigationListConfig;
+
+  private unlisteners: (() => void)[] = [];
+
+  @Input()
+  set configuration(config: Partial<KeyNavigationListConfig>) {
+    this.config = {
+      ...this.defaultConfig,
+      dir: this.host.nativeElement.getAttribute('rtl'),
+      ...config,
+    };
+  }
+
+  constructor(
+    private host: ElementRef<HTMLElement>,
+    private renderer: Renderer2,
+    @Inject(ClrListKeyNavigationConfigToken) private defaultConfig: KeyNavigationListConfig,
+    @Inject(FocusableElementsToken) private focusableElements: string[]
+  ) {
+    this.config = {
+      ...this.defaultConfig,
+      dir: this.host.nativeElement.getAttribute('rtl'),
+    };
+  }
+
+  ngAfterViewInit() {
+    this.initializeTabIndex();
+    this.unlisteners.push(this.renderer.listen(this.host.nativeElement, 'click', this.clickItem.bind(this)));
+    this.unlisteners.push(this.renderer.listen(this.host.nativeElement, 'keydown', this.focusItem.bind(this)));
+  }
+
+  ngOnDestroy() {
+    this.unlisteners.forEach(unlisten => unlisten());
+  }
+
+  private initializeTabIndex() {
+    if (this.config.manageFocus && this.config.manageTabindex) {
+      initializeKeyListItems(this.getListItems());
+    }
+  }
+
+  private clickItem(e: Event) {
+    const activeItem = this.getActiveItem(e);
+    if (activeItem) {
+      this.setActiveItem(e, activeItem);
+    }
+  }
+
+  private focusItem(e: KeyboardEvent) {
+    if (validKeyNavigationCode(e)) {
+      const activeItem = this.getActiveItem(e);
+      if (activeItem) {
+        const { next, previous } = getNextKeyListItem(activeItem, Array.from(this.getListItems()), {
+          ...this.config,
+          code: e.code as KeyNavigationCode,
+        });
+
+        if (next !== previous) {
+          this.setActiveItem(e, this.getListItems()[next], this.getListItems()[previous]);
+        }
+      }
+    }
+  }
+
+  private getActiveItem(e: Event) {
+    return Array.from(this.getListItems()).find(
+      c => c === (e.target as HTMLElement).closest(this.getListItems()[0].tagName.toLocaleLowerCase()) ?? c === e.target
```

Comment:
This code is unnecessarily expensive because it runs the DOM query several times. I recommend adding `const listItems = this.getListItems();` so that `getListItems` is only called once.

Reply:
Yea I thought about that as well. Will do.